graphql_compiler/compiler/emit_sql.py

```diff
@@ -107,6 +127,21 @@ def __init__(self, sql_schema_info, ir):
         self._outputs = []  # sqlalchemy Columns labelled correctly for output
         self._filters = []  # sqlalchemy Expressions to be used in the where clause
 
+        # The folded subquery being constructed as the IR is processed (reset for each fold)
+        self._fold_from_clause = None  # the main sqlalchemy Selectable for the fold subquery
+        self._fold_outputs = []  # sqlalchemy fold subquery outputs
+        self._fold_group_by = []  # sqlalchemy fold subquery Columns to GROUP BY
+
+        # Information needed to LEFT JOIN fold subquery to main Selectable (reset for each fold)
+        self._pre_fold_location = None  # the current Location prior to fold
+        self._fold_location = None  # the fold's FoldScopeLocation
+        self._fold_join_vertex_field = None  # field to perform LEFT JOIN on
+
+        # Information to keep track of fold aliases
+        self._folded_output_aliases = {}  # mapping FoldScopeLocations to their intermediate alias
+        self._folded_output_count = 1
+        self._fold_count = 1
```

Comment:
There's 9 mutable variables needed just for fold. This is fine in this draft PR, to demonstrate that `@fold` works. However, this is where most of the code complexity comes from, and can lead to bugs. Mutable state is the hardest to document, so let's have as little of it as possible. I think simplifying the state is a worthwhile investment, and it looks to me like it would be necessary to implement traversals inside fold. Some techniques you can try:
- Try to adopt the existing fields instead of defining new ones when possible. Is it necessary to store both `_current_location` and `_fold_location`? Do you have to update both of them when you're inside fold? Is it necessary to have both `_aliases` and `_folded_output_aliases`?
- Group some fields into a namedtuple, reducing visual clutter. Examples:
    - The query components like `_fold_from_clause`, `_fold_outputs` and `_fold_group_by` are all part of one query. You could create a `QueryBuilder` class or namedtuple that contains all parts of the query, which will simplify even the non-fold code that generates the final result.
    - The `_fold_count` and `_folded_output_count` are only used to generate unique names, but take up a central position in the complicated state we need to manage. You can pull out this functionality out of this class into a `UniqueAliasGenerator`, making these things invisible.
- Derive some of the fields from others, converting them into pure functions. See `self._current_classname` as an example.

There might be other ways that I'm not seeing. Feel free to play around and see what works.